Improving the UX of the conversation view.

misterPaul0 opened the pull request:
<!--
Thanks for contributing to the project!
Please help us keep this project in good shape by going through this checklist.
Replace the empty checkboxes [ ] below with checked ones [X] as they are completed
Remember, you can preview this before saving it.
-->

<!-- You can remove this first section if you have contributed before -->

### First time contributor checklist:

- [x] I have read the [README](https://github.com/signalapp/Signal-Desktop/blob/master/README.md) and [Contributor Guidelines](https://github.com/signalapp/Signal-Desktop/blob/master/CONTRIBUTING.md)
- [x] I have signed the [Contributor Licence Agreement](https://signal.org/cla/)

### Contributor checklist:

- [x] My contribution is **not** related to translations. _Please submit translation changes via our [Signal Desktop Transifex project](https://www.transifex.com/signalapp/signal-desktop/)._
- [x] My commits are in nice logical chunks with [good commit messages](http://chris.beams.io/posts/git-commit/)
- [x] My changes are [rebased](https://medium.freecodecamp.org/git-rebase-and-the-golden-rule-explained-70715eccc372) on the latest [`development`](https://github.com/signalapp/Signal-Desktop/tree/development) branch
- [x] A `yarn ready` run passes successfully ([more about tests here](https://github.com/signalapp/Signal-Desktop/blob/master/CONTRIBUTING.md#tests))
- [x] My changes are ready to be shipped to users

### Description

<!--
Describe briefly what your pull request changes. Focus on the value provided to users.

Does it address any outstanding issues in this project?
  https://github.com/signalapp/Signal-Desktop/issues?utf8=%E2%9C%93&q=is%3Aissue
  Reference an issue with the hash symbol: "#222"
  If you're fixing it, use something like "Fixes #222"

Please write a summary of your test approach:
  - What kind of manual testing did you do?
  - Did you write any new tests?
  - What operating systems did you test with? (please use specific versions: http://whatsmyos.com/)
  - What other devices did you test with? (other Desktop devices, Android, Android Simulator, iOS, iOS Simulator)
-->

Please check this screenshot to see the changes in action:

![output](https://user-images.githubusercontent.com/2820002/104548010-fb321500-5683-11eb-8cd8-193ef30db112.png)

Consecutive messages from the same author are now
 - displayed closer together, 
 - have their border radius' changed to look more congruent,
 - only have one profile picture / username per block of messages

In addition, the timestamp for the message is now displayed on the right of each message and does not increase the vertical height of the message unless there is insufficient space for it to be displayed at the end of the message.

These changes combined make for a significantly improved viewing experience for the user.

My test approach was to generate as many combinations of message sequences as I could think of to ensure different orders of messages would look correct. 

I needed to access conversation data from each message to achieve this, a process that appears difficult under the current architecture. For this reason I've used window.getConversations to find the conversation for the message and check the messages before/after it. If you have a suggestion for a better way to achieve this, please let me know!

tanujkucp commented:
The new UI changes you have done are very good and will help readability for users. 
Hope this is merged soon with some changes if needed.

psblah commented:
Nice, much nicer message bubbles

kaimast commented:
Thanks! I was about to file a bug report requesting this.

misterPaul0 commented:
I've updated the pull request to fix some styling issues:

- Metadata spacing fixed to allow for more space for older dates (which are longer) and for the "Sending Failed" message
- Link posts now style the same as the rest
- Fixed an issue with consecutive outgoing attachment messages being offset incorrectly
- Added a small 4px border radius instead of 0px to consecutive messages to be consistent with native apps.

GitMatze commented:
This would be a long-awaited feature! But there's still some work to be done. Here's a list of issues I found: 

- For large windows the time stamps of audio messages jump to the top. 
  ![image](https://user-images.githubusercontent.com/45139485/105099365-db4b8880-5aab-11eb-8c41-d3bebec94981.png)

- For small windows the time stamps touch the inner bubble 
 ![image](https://user-images.githubusercontent.com/45139485/105105241-fde19f80-5ab3-11eb-9b78-36fe94d9c25b.png)


- Quotes are not displayed correctly: 
![image](https://user-images.githubusercontent.com/45139485/105100059-04204d80-5aad-11eb-9692-61a3d080c046.png)

- The upper corners of link previews remain round. 
 ![image](https://user-images.githubusercontent.com/45139485/105100683-f0291b80-5aad-11eb-97d2-52a56a70814c.png)

- Time stamps and messages are too close sometimes.
![image](https://user-images.githubusercontent.com/45139485/105101180-c290a200-5aae-11eb-8b85-62ed70e3f8bd.png)

- Map previews also remain round, only the background seems to be affected
 ![image](https://user-images.githubusercontent.com/45139485/105103538-0be1f100-5ab1-11eb-8307-d08719be4438.png)

misterPaul0 commented:
Thank you @GitMatze those screenshots were very helpful. I've resolved the time stamp spacing issues on all situations you raised, the border radius on link posts, and the audio placement issues. How do you get the map preview to appear? I'm pretty new to this software. Also the quotes issue, how do I reproduce?

GitMatze commented:
You first have to create these messages with your regular Signal app, preferably with the smartphone app since the Desktop app lacks some features. Just use the Note to self chat for that. Then you start your regular Signal Desktop so that these messages are synced. Finally, you have to use this trick described in https://github.com/signalapp/Signal-Desktop/blob/master/CONTRIBUTING.md#the-staging-environment to transfer your real conversations to the development environment. 
To get these quote messages just press and hold an already sent message and respond to it by clicking the left arrow in the header bar. Let me know if you have further questions.
Note that there is a lot more to test here. Every type of message (text, audio, video, pic, link, location, file, contact, gif...) can be directly responded to using again each of these types. Definitely a lot of scenarios.

misterPaul0 commented:
I think the quote issue was resolved with the audio spacing fix? 

Just the map issue left to fix from that list I think. I'll try to recreate a map issue based on your description.

I agree there is still plenty of testing left to do, I'll do my best to QA but help like your list of bugs will give us the best chance of getting this merged. If the community can continue to help with testing different scenario's I'll try to fix them as fast as you can find them. 

I'm not super happy with my window.getConversations method of figuring out what messages are before/after but I can't figure out a better way to get this data, so any devs more knowledgeable of the codebase I'm happy to incorporate suggestions to improve the structure.

GitMatze commented:
I'd also love to see this feature getting merged. However I see a lot of little issues, which is why I think that this pull request is better split up into one dealing with the grouping and one dealing with the time stamps being placed more efficiently. The grouping feature should be much easier to solve and also it's the one that is clearly lacking compared to the mobile app. The time stamp feature would be an entirely new design choice for Signal.

Here's a list of issues. 
- This is still too cose 
![image](https://user-images.githubusercontent.com/45139485/105328705-9ec57d00-5bd0-11eb-91d1-3eee2b00b8a9.png)
- Actually, squeezing those two lines together is the wrong aproach, I think. Either they have to be put next to each other or, if there's not enough horizontal space, on top of each other but with proper line spacing. Like here: 
![image](https://user-images.githubusercontent.com/45139485/105331826-2fea2300-5bd4-11eb-9fff-c236350cb2e8.png)

- The time stamps of received audio messages are on the right, the ones of received text messages on the left
- The corners of the type of link previews where the picture is above still remain rounded 
![image](https://user-images.githubusercontent.com/45139485/105330562-bd2c7800-5bd2-11eb-9a41-2e56a183d3f8.png)
- The maps issue goes for sent fotos and GIFs as well:
![image](https://user-images.githubusercontent.com/45139485/105332660-2614ef80-5bd5-11eb-9b4b-ac3d227adfe5.png)
- responding to a text message with a foto leads to this: 
![image](https://user-images.githubusercontent.com/45139485/105335263-22cf3300-5bd8-11eb-9eae-3cdc6d7ce533.png)
- Responding to an image leads to this. Note that the upper left corner of the image is not sharp. In the mobile version it is. 
![image](https://user-images.githubusercontent.com/45139485/105335659-9cffb780-5bd8-11eb-9166-a9056128d2ee.png)

GitMatze commented:
https://github.com/signalapp/Signal-Desktop/pull/4346 is a very similar PR that was not followed up on. Might be helpful to look into this.

n-gao commented:
I really like these changes. But, I think the implementation could need some polishing. I am not a big fan of negative margins as they tend to difficult things. So, instead of moving the message out of its original position, it is better to adjust the original padding in `.module-timeline__message-container`. I got an implementation of this without any weird inconsitencies.
I would suggest you to take a look at it. Otherwise, if you want to I can open a PR. https://github.com/n-gao/Signal-Desktop-Retouched/tree/conversation_ux

GitMatze commented:
@n-gao In case you want to open a PR for this, note that for images, link previews, quotes etc there is another inner border, that I think does not yet behave correctly in your implementation. In the list above, that'd be bullet no 4 and the last one. Otherwise, it looks really good!

n-gao commented:
I caught the cases for images and quotes but missed links. Thanks for pointing that out! I fixed it for links as well. Are there any other attachments that may occur?

GitMatze commented:
I have a test conversation that should cover most scenarios and it looks perfect for all of them now. Good work!

n-gao commented:
I opened #4993

EvanHahn-Signal commented:
Sorry for the lack of updates here.

We have some intention of implementing this feature and have started to build the infrastructure to make it possible. I think we're going to implement this ourselves, but we will refer to your hard work (and the work of similar PRs) as a template.

I'm going to close this issue because we won't merge it as is, but again, thank you for your efforts.
